### PR TITLE
Retry requests on error.

### PR DESCRIPTION
**Pull request by ryy**

### CODE REVIEW

File: lib/embulk/input/twitter_ads_analytics.rb [hunk 1]
```diff
@@ -280,6 +317,20 @@ def entity_plural(entity)
...
+
+      def exp_backoff_sec(response:, retries:)
+        rate_limit_reset_timestamp = get_rate_limit_reset_timestamp(response: response)
+        (rate_limit_reset_timestamp.presence || (Time.zone.now + retries.second)).to_i - Time.zone.now.to_i
+      end
+
+      # https://developer.twitter.com/ja/docs/twitter-ads-api/rate-limiting
+      def get_rate_limit_reset_timestamp(response:)
+        # It is stored in the header in the format key => [value].
+        # @example x-rate-limit-reset"=>["1638316658"]
```
Comment (kekekenta) [inline lib/embulk/input/twitter_ads_analytics.rb:358]: `x-rate-limit-reset` returns a unix timestamp?
In that case, we should convert to seconds for wait.

File: lib/embulk/input/twitter_ads_analytics.rb [hunk 2]
```diff
@@ -280,6 +317,20 @@ def entity_plural(entity)
           "FUNDING_INSTRUMENTS"
         end
       end
+
+      private
+
+      def exp_backoff_sec(response:, retries:)
```
Comment (kekekenta) [inline lib/embulk/input/twitter_ads_analytics.rb]: This method is no longer an exponential backoff.

File: lib/embulk/input/twitter_ads_analytics.rb [hunk 3]
```diff
@@ -226,33 +229,67 @@ def get_access_token
...
+            metric_groups: @metric_groups.join(","),
+            start_time: chunked_time[:start_time],
+            end_time: chunked_time[:end_time],
+            placement: @placement,
+            granularity: @granularity,
+          }
+          response = access_token.request(:get, "https://ads-api.twitter.com/9/stats/accounts/#{@account_id}?#{URI.encode_www_form(params)}")
+          if response.code != "200"
+            Embulk.logger.error "#{response.body}"
+            raise
```
Comment (kekekenta) [inline lib/embulk/input/twitter_ads_analytics.rb]: Pleas define a concrete Standard Error class and raise/catch it.

File: lib/embulk/input/twitter_ads_analytics.rb [hunk 4]
```diff
@@ -280,6 +317,20 @@ def entity_plural(entity)
           "FUNDING_INSTRUMENTS"
         end
       end
+
+      private
+
+      def exp_backoff_sec(response:, retries:)
+        rate_limit_reset_timestamp = get_rate_limit_reset_timestamp(response: response)
+        (rate_limit_reset_timestamp.presence || (Time.zone.now + retries.second)).to_i - Time.zone.now.to_i
```
Comment (kekekenta) [inline lib/embulk/input/twitter_ads_analytics.rb]:
```suggestion
        if rate_limit_reset_timestamp.present?
          return [rate_limit_reset_timestamp - Time.now.to_i, 0].max
        else
          return retries.second
        end
```

File: lib/embulk/input/twitter_ads_analytics.rb (hunk 2 above)
Comment (kekekenta) [reply lib/embulk/input/twitter_ads_analytics.rb]: So, please rename the method name.

File: lib/embulk/input/twitter_ads_analytics.rb (hunk 4 above)
Comment (ryy) [reply lib/embulk/input/twitter_ads_analytics.rb]: fixed by https://github.com/trocco-io/embulk-input-twitter_ads_analytics/pull/6/commits/083cc9b46d6b7e2a3e1b52aa06b33a27b87229c9

File: lib/embulk/input/twitter_ads_analytics.rb (hunk 3 above)
Comment (ryy) [reply lib/embulk/input/twitter_ads_analytics.rb]: fixed by https://github.com/trocco-io/embulk-input-twitter_ads_analytics/pull/6/commits/25bd52add26f2b6e35ba737b4a7ab832a3ec9b21

File: lib/embulk/input/twitter_ads_analytics.rb (hunk 2 above)
Comment (ryy) [reply lib/embulk/input/twitter_ads_analytics.rb]: fixed by https://github.com/trocco-io/embulk-input-twitter_ads_analytics/pull/6/commits/255ad1528a4205f3139fc0063be2bc875be6f544

**Review by kekekenta — APPROVED** [review]
LGTM

**Review by d-hrs — APPROVED** [review]

**Pull request merged**